sklearn/utils/stats.py

```diff
@@ -1,11 +1,150 @@
+from collections.abc import Iterable
+
 import numpy as np
 
 from .extmath import stable_cumsum
 from .fixes import _take_along_axis
 
 
-def _weighted_percentile(array, sample_weight, percentile=50):
-    """Compute weighted percentile
+def _squeeze_arr(arr, n_dim):
+    return arr[0] if n_dim == 1 else arr
+
+
+def _nearest_rank(
+    array, sorted_idx, percentile, cum_weights, n_rows, n_cols, n_dim,
+):
+    """Compute weighted percentile using the nearest-rank approach.
+
+    Percentile can be computed using the nearest-rank:
+    https://en.wikipedia.org/wiki/Percentile
+
+    n = ceil(P / 100 * N)
+
+    where n is the index of the percentile value of the percentile P and N is
+    the number of samples.
+
+    This definition can be adapted with weights:
+
+    n_w = ceil(P / 100 * S_w) where S_w is the sum of the weights.
+    """
+    adjusted_percentile = percentile * cum_weights[-1]
+    percentile_idx = np.array([
+        np.searchsorted(cum_weights[:, i], adjusted_percentile[i])
+        for i in range(cum_weights.shape[1])
+    ])
+    percentile_idx = np.array(percentile_idx)
+    # In rare cases, percentile_idx equals to sorted_idx.shape[0]
+    max_idx = n_rows - 1
+    percentile_idx = np.apply_along_axis(
+        lambda x: np.clip(x, 0, max_idx), axis=0, arr=percentile_idx
+    )
+
+    col_index = np.arange(n_cols)
+    percentile_in_sorted = sorted_idx[percentile_idx, col_index]
+    percentile_value = array[percentile_in_sorted, col_index]
+    return _squeeze_arr(percentile_value, n_dim)
+
+
+def _interpolation_closest_ranks(
+    array, sorted_idx, sample_weight, sorted_weights, percentile, cum_weights,
+    interpolation, n_rows, n_cols, n_dim,
+):
+    """Compute the weighted percentile by interpolating between the adjacent
+    ranks.
+    Percentile can be computed with 3 different alternatives when using the
+    interpolation between adjacent ranks:
+    https://en.wikipedia.org/wiki/Percentile
+
+    These 3 alternatives depend of the value of a parameter C. NumPy uses
+    the variant where C=0 which allows to obtain a strictly monotonically
+    increasing function which is defined as:
+
+    P = (x - 1) / (N - 1),
+
+    where x in [1, N]
+
+    Weighted percentile change this formula by taking into account the
+    weights instead of the data frequency.
+
+    P_w = (x - w) / (S_w - w),
+
+    where x in [1, N], w being the weight and S_w being the sum of the weights.
+    """
```

Comment:
I understand the desire to be consistent with numpy implementation of `C=0` but I do think extrapolating to samples with weights causes problems. (copied over from old PR) It 'ignores' the last weight so, it doesn't give the same answer as with `np.percentile` and repeating the array by integer weights:
```
>>> _weighted_percentile(np.array([1, 2, 3, 4]), sample_weight=np.array([1, 1, 1, 5]),
...                      percentile=50, interpolation="linear")
3.3
>>> np.percentile(np.array([1, 2, 3, 4, 4, 4, 4, 4]), 50, interpolation='linear')
4.0
>>> np.median(np.array([1, 2, 3, 4, 4, 4, 4, 4]))
4.0
```
'ignoring' the last weight also means that these 2 have the same result:
```
>>> _weighted_percentile(np.array([1, 2, 3, 4]), sample_weight=np.array([1, 1, 1, 5]),
...                      percentile=50, interpolation="linear")
3.3
>>> _weighted_percentile(np.array([1, 2, 3, 4]), sample_weight=np.array([1, 1, 5, 5]),
...                      percentile=50, interpolation="linear")
3.3
```
It doesn't have the same properties when implemented with unit weights. I am not sure when it would be desirable to use this implementation.

Reply:
The repeating issue will not be a big deal because we are not going to use the linear interpolation to keep this property.

Reply:
Regarding the case

```python
>>> _weighted_percentile(np.array([1, 2, 3, 4]), sample_weight=np.array([1, 1, 1, 5]),
...                      percentile=50, interpolation="linear")
```
the percentile requested is between 3 and 4. It will be computed using linear interpolation. In this case, it is weird because we would have the intuition that the 4 should be more attractive. But then it means that linear interpolation is probably that we don't want with weighted percentile but we would expect a more non-linear interpolation.